### PR TITLE
fix(Core/Character): TeleportTo() in HandlePlayerLoginFromDB()

### PR DESCRIPTION
**Pull request by PkllonG**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/16307

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 1]
```diff
@@ -892,14 +892,18 @@ void WorldSession::HandlePlayerLoginFromDB(LoginQueryHolder const& holder)
...
-        AreaTriggerTeleport const* at = sObjectMgr->GetGoBackTrigger(pCurrChar->GetMapId());
-        if (at)
-            pCurrChar->TeleportTo(at->target_mapId, at->target_X, at->target_Y, at->target_Z, pCurrChar->GetOrientation());
-        else
-            pCurrChar->TeleportTo(pCurrChar->m_homebindMapId, pCurrChar->m_homebindX, pCurrChar->m_homebindY, pCurrChar->m_homebindZ, pCurrChar->GetOrientation());
-
-        // Probably a hackfix, but currently the best workaround to prevent character names showing as Unknown after teleport out from instances at login.
-        pCurrChar->GetSession()->SendNameQueryOpcode(pCurrChar->GetGUID());
+        ObjectGuid playerGUID = pCurrChar->GetGUID();
+        Map* map = pCurrChar->FindMap();
```
Comment (PkllonG) [inline src/server/game/Handlers/CharacterHandler.cpp:896]: Prevent crashes
```
        ObjectGuid playerGUID = pCurrChar->GetGUID();
        Map* map = pCurrChar->FindMap();
```

**Comment by PkllonG** [comment]
done

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 2]
```diff
@@ -892,14 +892,18 @@ void WorldSession::HandlePlayerLoginFromDB(LoginQueryHolder const& holder)
...
+        pCurrChar->m_Events.AddEventAtOffset([map, playerGUID]()
```
Comment (Kitzunu) [inline src/server/game/Handlers/CharacterHandler.cpp:897]: why

Comment (PkllonG) [reply src/server/game/Handlers/CharacterHandler.cpp:897]: This error can be avoided by delaying the removal of the player from the instance.

**Comment by walkline** [comment]
Have you tried to use `SetMustDelayTeleport`? Like this:

```
    SetMustDelayTeleport(true);
AreaTriggerTeleport const* at = sObjectMgr->GetGoBackTrigger(pCurrChar->GetMapId()); 
     if (at) 
         pCurrChar->TeleportTo(at->target_mapId, at->target_X, at->target_Y, at->target_Z, pCurrChar->GetOrientation()); 
     else 
         pCurrChar->TeleportTo(pCurrChar->m_homebindMapId, pCurrChar->m_homebindX, pCurrChar->m_homebindY, pCurrChar->m_homebindZ, pCurrChar->m_homebindO);
    SetMustDelayTeleport(false);
```

**Comment by PkllonG** [comment]
> Have you tried to use `SetMustDelayTeleport`? Like this:
> 
> ```
>     pCurrChar->SetMustDelayTeleport(true);
> AreaTriggerTeleport const* at = sObjectMgr->GetGoBackTrigger(pCurrChar->GetMapId()); 
>      if (at) 
>          pCurrChar->TeleportTo(at->target_mapId, at->target_X, at->target_Y, at->target_Z, pCurrChar->GetOrientation()); 
>      else 
>          pCurrChar->TeleportTo(pCurrChar->m_homebindMapId, pCurrChar->m_homebindX, pCurrChar->m_homebindY, pCurrChar->m_homebindZ, pCurrChar->m_homebindO);
>     pCurrChar->SetMustDelayTeleport(false);
> ```
I just tried SetMustDelayTeleport and it didn't fix the error.

**Comment by walkline** [comment]
It just looks a little hacky to me. I think the right way to handle it is to pass the correct coordinates at [this](https://github.com/azerothcore/azerothcore-wotlk/pull/19892/files#diff-596cc52dc80297752bc492b498987b3cfd7fd685bff2e8374bcec3e0b871fda6R815-R821) stage. 
However, that would likely require a lot of effort to make it work, right?

So, maybe we can use this approach.

But with this approach, you don't want to capture the map pointer. After 1 second, it could be invalid. 
So, use `ObjectAccessor::FindPlayer(playerGUID)` instead of `ObjectAccessor::GetPlayer(map, playerGUID)`.

**Pull request closed without merging**